great_expectations/zep/metadatasource.py

```diff
@@ -3,155 +3,89 @@
 
-        # TODO: extract asset type details
-
-        sources.register_factory(cls, _datasource_factory)
+        sources.register_factory(cls, _datasource_factory, asset_types=asset_types)
 
         return super().__new__(meta_cls, cls_name, bases, cls_dict)
 
-
-# class FileAsset:
-#     pass
-
-
-# class PandasDatasource(metaclass=MetaDatasouce):
-
-#     asset_types = [FileAsset]
-
-#     def __init__(self, name: str):
-#         self.name = name
-
-#     def __repr__(self):
-#         return f"{self.__class__.__name__}(name='{self.name}')"
-
-#     def add_csv(self, foo="foo", bar="bar", sep=","):
-#         """I'm a docstring!!"""
-#         # NOTE: should this return the datasource or the csv asset?
-#         return self
-
-
-# class TableAsset:
-#     pass
-
-
-# class PostgresDatasource(metaclass=MetaDatasouce):
-#     asset_types = [TableAsset]
-
-#     def __init__(self, name: str, connection_str: str):
-#         self.name = name
-#         self.connection_str = connection_str
-
-
-class DataContext:
-    """
-    NOTE: this is just a scaffold for exploring and iterating on our ZEP prototype
-    this will be formalized and tested prior to release.
-
-    Use `great_expectations.get_context()` for a real DataContext.
-    """
-
-    _context = None
-
     @classmethod
-    def get_context(cls) -> DataContext:
-        if not cls._context:
-            cls._context = DataContext()
-
-        return cls._context
-
-    def __init__(self) -> None:
-        self._sources: _SourceFactories = _SourceFactories()
-        print(f"4. Available Factories - {self._sources.factories}")
-
-    @property
-    def sources(self) -> _SourceFactories:
-        return self._sources
-
-
-def get_context() -> DataContext:
-    print("3. Getting context")
-    context = DataContext.get_context()
-    return context
-
-
-if __name__ == "__main__":
-    context = get_context()
-    # context.sources.add_pandas("taxi")
-    # context.sources.add_postgres("taxi2", connection_str="postgres://...")
+    def _inject_asset_method(
+        cls: Type[MetaDatasource],
+        ds_cls_dict: Dict[str, Callable],
+        asset_type: Type[DataAsset],
+    ) -> None:
+        LOGGER.info(f"1c. Injecting `add_<ASSET_TYPE>` method for {asset_type}")
+        method_name = f"add_{_get_simplified_name_from_type(asset_type)}"
+
+        method_already_defined = ds_cls_dict.get(method_name)
+        if method_already_defined:
+            LOGGER.info(
+                f"  {asset_type.__name__} method `{method_name}()` already defined"
+            )
+        attr_annotations = asset_type.__dict__.get("__annotations__", "")
+
+        # TODO: update signature with `attr_annotations`
+        def _add_asset(self: Datasource, name: str, *args, **kwargs):
+            LOGGER.info(f"6. Creating `{asset_type.__name__}` '{name}' ...")
+            data_asset = asset_type(name, *args, **kwargs)
+            self.assets[name] = data_asset
+            return data_asset
```

Comment:
Don't do this.
Datasources must define the asset-adding methods.

Reply:
Let's use a decorator to enforce the asset_method is added

Reply:
We are ejecting the `_add_asset` injection stuff for now.
We may revisit this in the future.

For future reference here is the last commit that contained this logic.
https://github.com/great-expectations/great_expectations/tree/ed7953d30ec7ed49296df4c8d1a2f32232d6b5a6/great_expectations/zep